### PR TITLE
Optimize Active Record batching

### PR DESCRIPTION
**Pull request by fatkodima**
Original `each_batch` implementation plucks specific ids and then uses them in `WHERE IN (...ids...)` query for the yielded batch.
This list is often very large (1000 by default), which can lead to inefficient queries and truncated sql queries in logs and monitoring tools. I'm sure everyone saw that monstrous queries of the half of the screen in size 😄 

This PR implements more efficient and compact queries by iterating in ranges (`WHERE id >= num1 AND id < num2`).

## Example

```ruby
User.in_batches(of: 3) do |relation|
  puts relation.to_sql; nil
end; nil
```

### Before

```sql
SELECT "users".* FROM "users" WHERE "users"."id" IN (1, 2, 3)
SELECT "users".* FROM "users" WHERE "users"."id" IN (4, 5, 6)
SELECT "users".* FROM "users" WHERE "users"."id" IN (7, 8, 9)
SELECT "users".* FROM "users" WHERE "users"."id" IN (10, 11, 12)
SELECT "users".* FROM "users" WHERE "users"."id" IN (13, 14)
```

### After

```sql
SELECT "users".* FROM "users" WHERE "users"."id" < 4 ORDER BY "users"."id" ASC LIMIT 3
SELECT "users".* FROM "users" WHERE "users"."id" >= 4 AND "users"."id" < 7 ORDER BY "users"."id" ASC LIMIT 3
SELECT "users".* FROM "users" WHERE "users"."id" >= 7 AND "users"."id" < 10 ORDER BY "users"."id" ASC LIMIT 3
SELECT "users".* FROM "users" WHERE "users"."id" >= 10 AND "users"."id" < 13 ORDER BY "users"."id" ASC LIMIT 3
SELECT "users".* FROM "users" WHERE "users"."id" >= 13 ORDER BY "users"."id" ASC LIMIT 3
```

This approach is already used in one of my gems (https://github.com/fatkodima/online_migrations/blob/master/lib/online_migrations/batch_iterator.rb) with great success.

After I implemented this patch, I noticed a similar PR - https://github.com/rails/rails/pull/42695, but seems like it is not working as advertized (still generates queries like in a "Before" approach) and my implementation looks simpler, imo. So I'm still opening this PR.

/cc @rafaelfranca @fxn as mentioned PR reviewers.

### CODE REVIEW

**Review by tenderlove — COMMENTED** [review]
This is an interesting idea, and I like it.  I'm worried about a few things though.  First, the order of records (as I commented).

Second, it looks like batches can be applied to arbitrary relations.  It seems (in theory) that the strategy implemented in this PR could over-select records.  It looks like this PR assumes that the ids will always be a consecutive and complete list (for example the ids will always be something like `[1, 2, 3, 4]` and never `[1, 3, 4]`).  Am I mistaken?

Third, how do we support or handle batching on AR objects that don't have integer primary keys?  For example, on PG you can use a UUID as a primary key, and sorting doesn't really make sense in that case.  I think we'd need to fall back to the old `IN` strategy in that case.

File: activerecord/lib/active_record/relation/batches.rb [hunk 1]
```diff
@@ -222,45 +222,53 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
...
         if load
-          records = batch_relation.records
-          ids = records.map(&:id)
-          yielded_relation = where(primary_key => ids)
-          yielded_relation.load_records(records)
+          records = batch_relation.uncached do
+            batch_relation.limit(batch_limit + 1).records
+          end
+
+          start = records[batch_limit]&.id
```
[the remark this answers is not in the document]
Comment (fatkodima) [reply activerecord/lib/active_record/relation/batches.rb:236]: Here we calculate `start` as from which id to start for the next iteration.

For collection of ids as [1, 2, 3, 5, 7, 20] and batch size as 3 and the first iteration:
- if `:order` is `:asc`, we get [1, 2, 3] and the `start` for the second iteration is 5
- if `:order` is `:desc`, we get [20, 7, 5] and the `start` for the second iteration is 3

**Comment by fatkodima** [comment]
@tenderlove Thanks for the review ❤️ 

> Second, it looks like batches can be applied to arbitrary relations. It seems (in theory) that the strategy implemented in this PR could over-select records. It looks like this PR assumes that the ids will always be a consecutive and complete list (for example the ids will always be something like [1, 2, 3, 4] and never [1, 3, 4]). Am I mistaken?

Ids can not be consecutive. For each iteration we select ids (by applying `LIMIT` clause) from the range `[x, y]`, where between x and y are `of` number items. 

> Third, how do we support or handle batching on AR objects that don't have integer primary keys? For example, on PG you can use a UUID as a primary key, and sorting doesn't really make sense in that case. I think we'd need to fall back to the old IN strategy in that case.

Batching is supported only for orderable primary keys (no uuids). And this is already mentioned in this method's documentation:
> This also means that this method only works when the primary key is orderable (e.g. an integer or string).

**Comment by fatkodima** [comment]
@byroot Can you please review this PR? 🙏 
I think Aaron has thousands of new notifications on GitHub and it would be impossible to get another round of reviews from him 😄

**Comment by byroot** [comment]
May take me a tiny while, as I need to dig in a few open question.

One quick one though:

```sql
SELECT "users".* FROM "users" WHERE "users"."id" < 4 ORDER BY "users"."id" ASC LIMIT 3
SELECT "users".* FROM "users" WHERE "users"."id" >= 4 AND "users"."id" < 7 ORDER BY "users"."id" ASC LIMIT 3
SELECT "users".* FROM "users" WHERE "users"."id" >= 7 AND "users"."id" < 10 ORDER BY "users"."id" ASC LIMIT 3
```

Why not:

```
SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT 3
SELECT "users".* FROM "users" WHERE "users"."id" > 3 ORDER BY "users"."id" ASC LIMIT 3
SELECT "users".* FROM "users" WHERE "users"."id" > 6 ORDER BY "users"."id" ASC LIMIT 3
```
?

**Comment by fatkodima** [comment]
We need to restrict on both ends, because it will not work correctly for cases like this: https://github.com/rails/rails/blob/66225cdb29f4bb262604b4f1e6d26a2ac3553795/activerecord/test/cases/batches_test.rb#L537-L546
when we apply conditions for the yielded relation. It can look for the records past the implicit end.

**Comment by byroot** [comment]
Yeah, I need to wrap my head around this PR and review it, but it keeps dropping from my hands...

cc @casperisfine (I'll try to do it tomorrow at work).

**Comment by casperisfine** [comment]
Ok, I finally understood what I was missing.

This is only for `Model.in_batches do |relation|`, not for `find_each / find_in_batches`. That makes much more sense now.

But then it opens the questions of why regular `find_each / find_in_batches` have to load all record upfront in certain cases, that could severly be optimized at well it seems. Anyway that's another question. I think i can review this one properly now.

File: activerecord/lib/active_record/relation/batches.rb [hunk 2]
```diff
@@ -222,45 +222,53 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
...
+
+          raise ArgumentError.new("Primary key not included in the custom select clause") unless records.first.id
+
+          batch_relation.load_records(records)
         else
-          ids = batch_relation.pluck(primary_key)
-          yielded_relation = where(primary_key => ids)
-        end
+          stop = batch_relation.uncached do
+            batch_relation.offset(batch_limit).pick(primary_key)
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/batches.rb:246]: So, if you do `offset(batch_limit -1)` you:

  - Scan 1 less record (not a huge deal but might as well)
  - No longer need the `inclusive: false` option.

I think that could simplify the code a bit

File: activerecord/lib/active_record/relation/batches.rb [hunk 3]
```diff
@@ -222,45 +222,53 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
...
+          end
 
-        break if ids.empty?
+          if stop
+            batch_relation = apply_finish_limit(batch_relation, stop, order, inclusive: false)
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/batches.rb:250]: I think we can use a range here:

```suggestion
            batch_relation = relation.where(primary_key => start..stop)
```

That should significantly simplify the code.

**Comment by fatkodima** [comment]
The original code hasn't `skip_query_cache!` on the `yielded_relation`. Should we have it? I does not make a lot of sense to cache query results called on that relation.

**Comment by byroot** [comment]
> I does not make a lot of sense to cache query results called on that relation.

Agreed.

**Comment by byroot** [comment]
Wait, I said the above assuming one thing:

  - If you do a write, even if you have `uncached` or `sip_query_cache` it should still flush the cache (I think that's the case but worth double checking)

File: activerecord/lib/active_record/relation/batches.rb (hunk 2 above)
Comment (fatkodima) [reply activerecord/lib/active_record/relation/batches.rb:246]: On an a batch edges, we need to have in a previous batch `id < stop` and in the following `id >= stop`. At the beginning of the loop we have `batch_relation = apply_start_limit(relation, start, order) if start` (which does `>=`), so we need to do `<` here and need to get a `batch_limit + 1`th record.
If we will do `offset(batch_limit - 1)`, we will have a batch of smaller size by 1.

File: activerecord/lib/active_record/relation/batches.rb (hunk 3 above)
Comment (fatkodima) [reply activerecord/lib/active_record/relation/batches.rb:250]: Unfortunately, we can't use ranges here, because it generates `id BETWEEN x AND y` sql query, which is equivalent to `id >= x AND id <= y`. When we batching in a `desc` order, `y <= x` and `BETWEEN` would return no results (for example, `x = 10`, `y = 1` - `id >= 10 AND id <= 1` makes no sense).

**Comment by fatkodima** [comment]
@byroot
>  I does not make a lot of sense to cache query results called on that relation

Seems like we should not disable query caching here https://github.com/rails/rails/blob/c00c07b66e5645b4941cbda2c4457300a7a10755/activerecord/test/cases/batches_test.rb#L717-L726. I rechecked, write queries already do flushing.

File: activerecord/lib/active_record/relation/batches.rb (hunk 3 above)
Comment (byroot) [reply activerecord/lib/active_record/relation/batches.rb:250]:
```ruby
order == :asc ? start..stop : stop..start
```

no?

**Comment by byroot** [comment]
> Seems like we should not disable query caching here

Well, OK then. If there's a test for it.

File: activerecord/lib/active_record/relation/batches.rb (hunk 3 above)
Comment (fatkodima) [reply activerecord/lib/active_record/relation/batches.rb:250]: This works, but we already have `apply_finish_limit`, used in other place, so this might be confusing why we do not use it.

**Comment by nvasilevski** [comment]
Hey folks, let me bring one use-case which I'm not exactly sure about
I think the change makes total sense when an empty relation being iterated over. However, I think that for a relation with some conditions this change may actually harm the performance.
Let's consider the following case:
```ruby
# having a ClothingItem model with schema like

  create_table :clothing_items, force: true do |t|
    t.string :color
    t.string :item_type
    t.string :size

    t.index [:item_type, :color, :size], unique: true
  end
```
Iteration like:

`ClothingItem.where(item_type: "t-shirt").in_batches(of: 2) {|b| puts b.to_sql}`
Would previously have resulted  in sql queries like 
```sql
SELECT "clothing_items".* FROM "clothing_items" WHERE "clothing_items"."item_type" = 't-shirt' AND "clothing_items"."id" IN (9, 10)
```

which is very efficient, and according to the sqlite `explain query plan` - `SEARCH TABLE clothing_items USING INTEGER PRIMARY KEY (rowid=?)`


However with this change the iteration is going to result in sql queries like:
```sql
SELECT "clothing_items".* FROM "clothing_items" WHERE "clothing_items"."item_type" = 't-shirt' AND "clothing_items"."id" >= 9 AND "clothing_items"."id" < 11 ORDER BY "clothing_items"."id" ASC LIMIT 2
```
which is less efficient because of the lack of index on the `t-shirt` column and according to `explain query plan`:
```
SEARCH TABLE clothing_items USING COVERING INDEX index_clothing_items_on_item_type_and_color_and_size (item_type=?)"
USE TEMP B-TREE FOR ORDER BY
```

And the `USE TEMP B-TREE ` part concerns me.

As I mentioned, we could add `t.index :item_type` to make the query efficient again, but I feel like it might be a little unexpected for an application owner to do that after an upgrade and it might not always be suitable to add an additional index. So I was wondering, perhaps we should make the change more visible trough changelog or even provide an option to choose one of two batching options?

Here is a reproduction script we can use:
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
  gem "debug"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :clothing_items, force: true do |t|
    t.string :color
    t.string :item_type
    t.string :size

    t.index [:item_type, :color, :size], unique: true
  end
end

class ClothingItem < ActiveRecord::Base
end



class BugTest < Minitest::Test
  def test_association_stuff
    ClothingItem.create(item_type: "t-shirt", color: "green", size: "s")
    ClothingItem.create(item_type: "t-shirt", color: "green", size: "m")
    ClothingItem.create(item_type: "t-shirt", color: "green", size: "l")
    ClothingItem.create(item_type: "t-shirt", color: "green", size: "xl")

    ClothingItem.create(item_type: "t-shirt", color: "red", size: "s")
    ClothingItem.create(item_type: "t-shirt", color: "red", size: "m")
    ClothingItem.create(item_type: "t-shirt", color: "red", size: "l")
    ClothingItem.create(item_type: "t-shirt", color: "red", size: "xl")

    ClothingItem.create(item_type: "t-shirt", color: "blue", size: "s")
    ClothingItem.create(item_type: "t-shirt", color: "blue", size: "m")
    ClothingItem.create(item_type: "t-shirt", color: "blue", size: "l")
    ClothingItem.create(item_type: "t-shirt", color: "blue", size: "xl")

    ClothingItem.create(item_type: "pants", color: "blue", size: "s")
    ClothingItem.create(item_type: "pants", color: "blue", size: "m")
    ClothingItem.create(item_type: "pants", color: "blue", size: "l")
    ClothingItem.create(item_type: "pants", color: "blue", size: "xl")

    ClothingItem.where(item_type: "t-shirt").in_batches(of: 2) do |batch|
      puts ClothingItem.connection.execute("explain query plan #{batch.to_sql}").to_a
    end
  end
end
```

**Comment by byroot** [comment]
That's an interesting point Nikita 🤔 .

What I'm not sure of, is whether the `IN ()` actually returned sorted records. I need to double check but I'm pretty sure that's not the case on MySQL at least.

**Comment by nvasilevski** [comment]
>If in_batches never guaranteed order, we can just continue to do so.

Not sure whether it was or should be guaranteed, but in my experience I'd say that for the most common use-case the order shouldn't really matter as I would expect `in_batches` to be used like:
```ruby
 User.where("last_login < ?", 3.days.ago).in_batches(of: 100) { |b| b.update_all(status: "deactivated") }
```
and the order shouldn't really matter
Similarly to an example like:
```ruby
 User.where("last_login < ?", 3.days.ago).in_batches(of: 100) do |b|
   user_ids = b.pluck(:id)
  Comment.where(user_id: user_ids).delete_all
end
```

Perhaps ordering could be optional and disabled by default for cases where batch should be ordered but I would expect that most use-cases shouldn't need it

**Comment by byroot** [comment]
> Perhaps ordering could be optional

Well, it kind already is as demonstrated above. If you want it you can do:

 User.where("last_login < ?", 3.days.ago).in_batches(of: 100) do |b|
  b.order(...).do_stuff
end
```

**Comment by nvasilevski** [comment]
>Well, it kind already is

You mean, prior to the change the order wasn't guaranteed? Because with this change, the `ORDER BY "clothing_items"."id" ASC LIMIT 2` part is added by default and if I want to keep my batching efficient, I'd have to do something like:
```ruby
  ClothingItem.where(item_type: "t-shirt").in_batches(of: 2) do |batch|
    batch.unscope(:order).do_stuff
  end
``` 

But even if we unscope the order, there will be a difference in query plans between the new query:
```
SELECT "clothing_items".* FROM "clothing_items" WHERE "clothing_items"."item_type" = 't-shirt' AND "clothing_items"."id" >= 9 AND "clothing_items"."id" < 11 
```

and the old one

```
SELECT "clothing_items".* FROM "clothing_items" WHERE "clothing_items"."item_type" = 't-shirt' AND "clothing_items"."id" IN (9, 10)
```

Out of curiosity, I'm going to try benchmarking these two queries in mysql

**Comment by nvasilevski** [comment]
>there will be a difference in query plans between the new query:

Nvm my comment, it seems like in mysql both queries result in the same execution plan (i guess sqlite confused me), which uses `PRIMARY` index to fetch data and `Using where` to filter by the `item_type` 
To be fair, this means in this particular case in mysql even `order by id` won't impact the performance as it's going to be ignored by the sever as `PRIMARY` index already sorts records.

Which means that in this case the new logic should be efficient as we don't send the large `ID IN (...)` sql to the server

And maybe my concern was sqlite specific and not very reasonable as I was confused by sqlite using `USING COVERING INDEX index_clothing_items_on_item_type_and_color_and_size` 😕 
In other words,
```sql
SELECT "clothing_items".* FROM "clothing_items" WHERE "clothing_items"."item_type" = 't-shirt' AND "clothing_items"."id" >= 9 AND "clothing_items"."id" < 11 ORDER BY "clothing_items"."id" ASC LIMIT 2
# and
SELECT "clothing_items".* FROM "clothing_items" WHERE "clothing_items"."item_type" = 't-shirt' AND "clothing_items"."id" IN (9, 10)
```
should be equally efficient from query-execution plan point of view as long as `PRIMARY` index is used to perform the query.

**Comment by etiennebarrie** [comment]
The execution plan can vary a lot based on runtime information though, I don't think we can infer a lot from small datasets.

Do we have benchmarks or examples of the inefficiency of the previous approach? If the only issue is the size of the query, that can easily be fixed by changing the value of the `of` parameter (maybe the default value). If the issue is logs, maybe that ought to be fixed instead.

The big problem I see with this approach, is similar to @nvasilevski's point, previously the main query included all the conditions, but then the batch relation was very simple, and all the batch queries were straight-forward. If you have outer joins, multiple levels of joins and conditions, now your `update_all` or `delete_all` is also going to get those, not to mention if you have multiple of these in the `in_batches` block.

**Comment by byroot** [comment]
According to https://stackoverflow.com/questions/3308280/is-there-a-performance-difference-between-between-and-in-with-mysql-or-in-sql-in `BETWEEN` (which is mostly a macro for `>= x AND <= y`) is faster than `IN`.

I tested it in prod on a large dataset with 1k ids, and it's indeed noticeably faster.

**Comment by byroot** [comment]
> previously the main query included all the conditions, but then the batch relation was very simple

Ok, I missed this part. Maybe this optimization only make sense when the `ids` are contiguous?

**Comment by byroot** [comment]
> Maybe this optimization only make sense when the `ids` are contiguous?

But that's probably rare in actual cases. 

@fatkodima I'm tempted to revert, at least for now, what do you think?

**Comment by fatkodima** [comment]
Please, wait until tomorrow. I will look into all the comments here at today's evening.

**Comment by byroot** [comment]
There's no rush. Another option might be to make this behavior opt-in.

**Comment by fatkodima** [comment]
Thank you guys for comments ❤️ Why you did not comment when this PR was open? 😄 

So, tested different queries on 2 tables of size 10 millions records each. For different queries got different results: sometimes the old approach was faster (by `3x`, for example), sometimes the new approach was faster (by `3x`, `2x`), sometimes they where ~.

The new approach is always 100% faster (unless I am missing something) only when iterating over the whole tables (where there are no conditions/joins/etc), like `User.in_batches/User.find_each/User.find_in_batches`. I got `155s` vs `25s` for this case. 

The pattern of iterating over the whole table is quite popular, for example:
1. when you need to index/reindex the whole table to the search engine
2. populate rows with some value when adding a new column
3. periodically iterate over the whole table to ensure all rows are valid (helps to detect bugs in the app)
4. export to CSV
5. iterate over the whole table to update all records by something
6. etc

If you think it is worth to optimize for such a use case, I will reword this PR with it. Otherwise, it should be just reverted.

**Comment by casperisfine** [comment]
> If you think it is worth to optimize for such a use case, I will reword this PR with it. Otherwise, it should be just reverted.

I'd say let's revert for now, and please open a new PR that makes it an option (`in_batches(something: true)`).

This way we can take all the time we want to consider the feature.

**Comment by simi** [comment]
> There's no rush. Another option might be to make this behavior opt-in.

I was thinking for some time how to contribute PostgreSQL cursor and also custom non-blocking iterator (since the deadlock within Fiber problem starting Ruby 3.0.2). Making adapter for custom iterating of batches would be great.

**Comment by fatkodima** [comment]
@byroot I am wondering - won't this be a not so good UX? 

Using this option, the user should know the internals of the batching method, he also probably needs to test both options on production or large enough db before the decision. But even when he enables or disables this config, it is possible when the size of the db increased or the schema changed, this config will start harming the performance. 

Wdyt about just testing if the user iterates over the whole table (no conditions/joins/etc) and implicitly use batching using ranges? And this approach makes the most sense only for the whole tables.

**Comment by byroot** [comment]
> Wdyt about just testing if the user iterates over the whole table (no conditions/joins/etc) and implicitly use batching using ranges?

If the implementation is simple enough, then 👍. e.g. if it's something super simple like `if self == klass.all` then that's great.

**Comment by nvasilevski** [comment]
Was also thinking about having several "strategies" and making `in_batches` "smart" in a way to use the ranges `id >= and id` strategy for relations with no conditions 👍

**Comment by fatkodima** [comment]
> `if self == klass.all`

This won't be as simple as this, because we have at least `default_scope`s 😄 I will try to implement, if it will be simple enough, I will open a PR.
